internal/provisioner/thanos.go

```diff
@@ -0,0 +1,218 @@
+		t.logger.Debugln("Main CNAME was already provisioned for thanos")
+	} else {
+		t.logger.Debugln("Main CNAME was not provisioned for thanos")
+		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(120)*time.Second)
+		defer cancel()
+
+		endpoint, err := getPrivateLoadBalancerEndpoint(ctx, "nginx", logger.WithField("thanos-action", "create"), t.kops.GetKubeConfigPath())
+		if err != nil {
+			return errors.Wrap(err, "couldn't get the load balancer endpoint (nginx) for Thanos")
+		}
+
+		logger.Infof("Registering DNS %s for %s", dns, app)
+		err = t.awsClient.CreatePrivateCNAME(dns, []string{endpoint}, logger.WithField("thanos-dns-create", dns))
+		if err != nil {
+			return errors.Wrap(err, "failed to create a CNAME to point to Thanos")
+		}
+	}
+
+	grpcDNS := fmt.Sprintf("%s-grpc.%s.%s", t.cluster.ID, app, privateDomainName)
+	if t.awsClient.IsProvisionedPrivateCNAME(grpcDNS, t.logger) {
+		t.logger.Debugln("GRPC CNAME was already provisioned for thanos")
+	} else {
+		t.logger.Debugln("GRPC CNAME was not provisioned for thanos")
+		ctx, cancel := context.WithTimeout(context.Background(), time.Duration(120)*time.Second)
+		defer cancel()
+
+		endpoint, err := getPrivateLoadBalancerEndpoint(ctx, "prometheus", logger.WithField("thanos-action", "create"), t.kops.GetKubeConfigPath())
+		if err != nil {
+			return errors.Wrap(err, "couldn't get the load balancer endpoint for Thanos")
+		}
+
+		logger.Infof("Registering GRPC DNS %s for %s", grpcDNS, app)
+		err = t.awsClient.CreatePrivateCNAME(grpcDNS, []string{endpoint}, logger.WithField("thanos-dns-create", grpcDNS))
+		if err != nil {
+			return errors.Wrap(err, "failed to create a CNAME to point to Thanos GRPC")
+		}
+	}
+	return nil
+}
+
+func (t *thanos) Destroy() error {
+	logger := t.logger.WithField("prometheus-action", "destroy")
+
+	privateDomainName, err := t.awsClient.GetPrivateZoneDomainName(logger)
+	if err != nil {
+		return errors.Wrap(err, "unable to lookup private zone name")
+	}
+	app := "thanos"
+	dns := fmt.Sprintf("%s.%s.%s", t.cluster.ID, app, privateDomainName)
+
+	logger.Infof("Deleting Route53 DNS Record for %s", app)
+	err = t.awsClient.DeletePrivateCNAME(dns, logger.WithField("thanos-dns-delete", dns))
+	if err != nil {
+		return errors.Wrap(err, "failed to delete Route53 DNS record")
+	}
+
+	grpcDNS := fmt.Sprintf("%s-grpc.%s.%s", t.cluster.ID, app, privateDomainName)
+	logger.Infof("Deleting GRPC Route53 DNS Record for %s", app)
+	err = t.awsClient.DeletePrivateCNAME(grpcDNS, logger.WithField("thanos-dns-delete", grpcDNS))
+	if err != nil {
+		return errors.Wrap(err, "failed to delete GRPC Route53 DNS record")
+	}
+
+	t.actualVersion = ""
+	return nil
+}
+
+func (t *thanos) Migrate() error {
+	return nil
+}
+
+func (t *thanos) NewHelmDeployment() *helmDeployment {
+	privateDomainName, err := t.awsClient.GetPrivateZoneDomainName(t.logger)
+	if err != nil {
+		t.logger.WithError(err).Error("unable to lookup private zone name")
```

Comment:
We store this value even if there is an error. I see in the `CreateOrUpgrade` method we look up the privateDomainName again. Do we need to even look it up here?

Reply:
@gabrieljackson Thanks you are right on that. We were doing the same thing in Prometheus, which we shouldn't. I updated both Prometheus and Thanos to pass the DNS as part of the NewHelmDeployment function. This way if there is an error we will catch it in the CreateORUpgrade where the privateDomainname is set.